pkg/controller/azvolume.go

```diff
@@ -0,0 +1,391 @@
+	}
+
+	//azVolume creation
+	if azVolume.Status == nil {
+		if err := r.TriggerCreate(ctx, azVolume.Name); err != nil {
+			klog.Errorf("failed to create AzVolume (%s): %v", azVolume.Name, err)
+			return reconcile.Result{Requeue: true}, err
+		}
+
+	} else if !azVolume.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err := r.TriggerDelete(ctx, azVolume.Name); err != nil {
+			//If delete failed, requeue request
+			return reconcile.Result{Requeue: true}, err
+		}
+	}
+
+	return reconcile.Result{Requeue: true}, nil
+}
+
+func (r *reconcileAzVolume) TriggerCreate(ctx context.Context, volumeName string) error {
+	var azVolume v1alpha1.AzVolume
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: volumeName}, &azVolume); err != nil {
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+
+	//Register finalizer
+	if err := r.InitializeMeta(ctx, volumeName); err != nil {
+		return err
+	}
+
+	if err := r.CreateAzVolume(ctx, &azVolume); err != nil {
+		klog.Errorf("failed to create volume %s: %v", azVolume.Spec.UnderlyingVolume, err)
+		return err
+	}
+
+	// Update status of the object
+	if err := r.UpdateStatus(ctx, azVolume.Name, false); err != nil {
+		return err
+	}
+	klog.Infof("successfully created volume (%s)and update status of AzVolume (%s)", azVolume.Spec.UnderlyingVolume, azVolume.Name)
+	return nil
+
+}
+
+func (r *reconcileAzVolume) TriggerDelete(ctx context.Context, volumeName string) error {
+	var azVolume v1alpha1.AzVolume
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: volumeName}, &azVolume); err != nil {
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+
+	if err := r.DeleteAzVolume(ctx, &azVolume); err != nil {
+		klog.Errorf("failed to delete volume %s: %v", azVolume.Spec.UnderlyingVolume, err)
+		return err
+	}
+
+	// Update status of the object
+	if err := r.UpdateStatus(ctx, azVolume.Name, true); err != nil {
+		return err
+	}
+	klog.Infof("successfully deleted volume (%s)and update status of AzVolume (%s)", azVolume.Spec.UnderlyingVolume, azVolume.Name)
+	return nil
+}
+
+func (r *reconcileAzVolume) UpdateStatus(ctx context.Context, volumeName string, isDeleted bool) error {
+	var azVolume v1alpha1.AzVolume
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: volumeName}, &azVolume); err != nil {
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+
+	if isDeleted {
+		if err := r.DeleteFinalizer(ctx, volumeName); err != nil {
+			klog.Errorf("failed to delete finalizer %s for azVolume %s: %v", AzVolumeFinalizer, azVolume.Name, err)
+			return err
+		}
+	}
+
+	updated := azVolume.DeepCopy()
+	if updated.Status == nil {
+		updated.Status = &v1alpha1.AzVolumeStatus{}
+	}
+	updated.Status.UnderlyingVolume = azVolume.Spec.UnderlyingVolume
+
+	if err := r.client.Status().Update(ctx, updated, &client.UpdateOptions{}); err != nil {
```

Comment:
For when `isDeleted == true`, once the finalizer is deleted, the object will be deleted, and this update call will likely return error and output unnecessary error messages.

Reply:
I believe lines 153-159 will prevent us from reaching here is `isDeleted == true`.

Reply:
Yup, commit https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/pull/20/commits/280828707ecc5e1e6bfb4572a63efba25996518b fixed what I mentioned in the comment :)